phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala

```diff
@@ -133,23 +142,28 @@ private final class ComRun(jettyClassLoader: ClassLoader,
   }
 
   private def receiveFrag(frag: String): Unit = synchronized {
-    /* The fragments are accumulated in an instance-wide buffer in case
-     * receiving a non-first fragment times out.
+    /* If the promise has already been completed, we cannot deliver new
+     * messages.
      */
-    fragmentsBuf ++= frag.substring(1)
-
-    frag.charAt(0) match {
-      case '0' =>
-        // Last fragment of a message, send it
-        val result = fragmentsBuf.result()
-        fragmentsBuf.clear()
-        onMessage(result)
-
-      case '1' =>
-        // There are more fragments to come; do nothing
-
-      case _ =>
-        throw new AssertionError("Bad fragmentation flag in " + frag)
+    if (!promise.isCompleted) {
```

Comment:
This has a risk of cutting of trialing messages. For example imagine the code:

```javascript
send("a");
send("b");
send("c");
send("d");
send("e");
exit();
```

If the JVM side drops messages after the run is completed, it might only call `onMessage` for say `a-d`.

Reply:
That risk also exists if we wait for the connection to be closed, AFAICT, because when the script terminates, it can hard-cut the WebSocket connection, and messages could be lost as well.

I first tried to wait for the connection to be closed, but the `onClose()` event was not reliable *at all*, so I would often wait indefinitely. It seems that `onClose()` is in high danger of not being called if `mgr.stop()` was already called.

Reply:
Ugh :( That is awful. Fair enough. Maybe drop a comment about this. IIUC this also does not cleanly satisfy the no `onMessage` in flight anymore when `future` is completed. But maybe we can live with that for now to unblock everything.

Reply:
I'm still trying something with a new `mgr.close()` that calls `webSocketConn.close()` internally, see if that works better. Otherwise I'll write a comment, yes.

Reply:
SG

Reply:
OK I think I've found a solution, in the latest commit.